src/resource/link/link.service.ts

```diff
@@ -0,0 +1,129 @@
+  constructor(
+    private readonly configService: ConfigService,
+    private readonly prisma: PrismaService,
+    private readonly httpService: HttpService,
+  ) {}
+
+  async create({ userId }: CreateLinkInput): Promise<CreateLinkResponse> {
+    const newLink = await this.prisma.link.create({
+      data: {
+        isDraft: true,
+        isVisible: true,
+        title: null,
+        url: null,
+        logoUrl: null,
+        user: { connect: { id: userId } },
+        createdAt: new Date().toISOString(),
+        updatedAt: new Date().toISOString(),
+      },
+    });
+
+    this.logger.log(`Created a new link for user ${userId}`);
+    return { id: newLink.id };
+  }
+
+  async update({ id, isVisible, title, url }: UpdateLinkInput) {
+    const linkToBeUpdated = await this.findLinkById(id);
+
+    const updatedData = {
+      isDraft: true,
+      isVisible,
+      title: _.isEmpty(title) ? linkToBeUpdated.title : title,
+      url,
+      logoUrl: linkToBeUpdated.logoUrl,
+    };
+
+    // Validate url if it's different to the existing url.
+    if (!_.isEqual(url, linkToBeUpdated.url) && !_.isEmpty(url)) {
+      const {
+        title,
+        site: { logo },
+      } = await this.validateUrl(url);
+
+      // Use the extracted title if the user did not specify one.
+      if (_.isEmpty(updatedData.title)) {
+        updatedData.title = title;
+      }
+      updatedData.logoUrl = logo;
+    }
+
+    // Check if link can be published
+    updatedData.isDraft =
+      _.isEmpty(updatedData.title) || _.isEmpty(updatedData.url) ? true : false;
+
+    const updatedLink = await this.prisma.link.update({
+      where: { id },
+      data: {
+        ...updatedData,
+        updatedAt: new Date().toISOString(),
+      },
+    });
+
+    if (updatedLink) {
+      this.logger.log(`Updated link ${id}.`);
+      return updatedLink;
+    } else {
+      this.logger.error(`Errored when updating link ${id}.`);
+    }
+  }
+
+  async findLinkById(id: string) {
+    this.logger.log(`Attempting to find link ${id}`);
+    const link = await this.prisma.link.findUnique({ where: { id } });
+
+    if (_.isNil(link)) {
+      this.logger.error(`Cannot find link ${id}`);
+    } else {
+      return link;
+    }
+  }
+
+  async validateUrl(url: string) {
+    this.logger.log(`Validating url ${url} received..`);
+    const encodedUrl = encodeURIComponent(url);
+    const { data, status } = await this.httpService
+      .get<UrlMetaResponse>(`${UrlMeta.API_URL}/?url=${encodedUrl}`, {
+        headers: {
+          Authorization: Buffer.from(
+            this.configService.get(UrlMeta.URL_META_AUTH_STRING),
+          ).toString('base64'),
+        },
+      })
+      .toPromise();
+
+    if (status !== HttpStatus.OK)
+      this.logger.error('Request errored with Url Meta API');
+
+    const { result, meta } = data;
+
+    if (result.status === UrlMeta.RESULT_ERROR) {
+      this.logger.error(`Failed to validate url. [Message: ${result.reason}]`);
```

Comment:
Now that you're documenting an error here, why don't we also throw an exception?

Reply:
At this stage, `logger.error` is extended to throw `BadRequestException` as well. 

The reason why it's implemented this way is if we just `throw new BadRequestException("message")`, the build-in exception handler will not log it. (__Yea, not because I haven't set up a custom exception handler, lol__).
<img width="990" alt="image" src="https://user-images.githubusercontent.com/42906455/173188966-bc177698-d688-4ea5-8b84-38f94c04198c.png">

So I extended `logger.error` so we can capture it.
<img width="921" alt="image" src="https://user-images.githubusercontent.com/42906455/173189125-1002f9f7-4430-4a66-9032-8289e96415e2.png">